QRCoder/QRCodeGenerator.cs

```diff
@@ -894,35 +897,69 @@ private static string ConvertToIso8859(string value, string Iso = "ISO-8859-2")
         /// </remarks>
         private static BitArray PlainTextToBinaryByte(string plainText, EciMode eciMode, bool utf8BOM, bool forceUtf8)
         {
-            byte[] codeBytes;
+            Encoding targetEncoding;
 
             // Check if the text is valid ISO-8859-1 and UTF-8 is not forced, then encode using ISO-8859-1.
             if (IsValidISO(plainText) && !forceUtf8)
-                codeBytes = Encoding.GetEncoding("ISO-8859-1").GetBytes(plainText);
+            {
+                targetEncoding = _iso8859_1;
+                utf8BOM = false;
+            }
             else
             {
                 // Determine the encoding based on the specified ECI mode.
                 switch (eciMode)
                 {
                     case EciMode.Iso8859_1:
                         // Convert text to ISO-8859-1 and encode.
-                        codeBytes = Encoding.GetEncoding("ISO-8859-1").GetBytes(ConvertToIso8859(plainText, "ISO-8859-1"));
+                        targetEncoding = _iso8859_1;
+                        utf8BOM = false;
                         break;
                     case EciMode.Iso8859_2:
+                        // Note: ISO-8859-2 is not natively supported on .NET Core
+                        //
+                        // Users must install the System.Text.Encoding.CodePages package and call Encoding.RegisterProvider(CodePagesEncodingProvider.Instance)
+                        // before using this encoding mode.
+                        if (_iso8859_2 == null)
+                            _iso8859_2 = Encoding.GetEncoding(28592); // ISO-8859-2
                         // Convert text to ISO-8859-2 and encode.
-                        codeBytes = Encoding.GetEncoding("ISO-8859-2").GetBytes(ConvertToIso8859(plainText, "ISO-8859-2"));
+                        targetEncoding = _iso8859_2;
+                        utf8BOM = false;
                         break;
                     case EciMode.Default:
                     case EciMode.Utf8:
                     default:
                         // Handle UTF-8 encoding, optionally adding a BOM if specified.
-                        codeBytes = utf8BOM ? Encoding.UTF8.GetPreamble().Concat(Encoding.UTF8.GetBytes(plainText)).ToArray() : Encoding.UTF8.GetBytes(plainText);
+                        targetEncoding = Encoding.UTF8;
                         break;
                 }
             }
 
+#if NETCOREAPP2_1_OR_GREATER || NETSTANDARD2_1
+            // We can use stackalloc for small arrays to prevent heap allocations
+            // Note that all QR codes should fit within 3000 bytes, so this code should never trigger a heap allocation unless an exception will be thrown anyway.
+            int count = targetEncoding.GetByteCount(plainText);
+            Span<byte> codeBytes = count < 3000 ? stackalloc byte[count] : new byte[count];
```

Comment:
The constant should be a multiple of 2, in order to create better machine code.

The stack-space must be zoroed*, and cpu registers, etc. are all multiple of 2.

\* except `[SkipLocalsInit]` is applied, then the zeroing can be omitted by the JIT
Note: if this attribute isn't available at the target framework, one can just provide a simple shim for it and the JIT will still recognize it.

Reply:
I'm not sure how this applies.  `3000` is used as a threshold to switch between `stackalloc` and `new byte[]`, and as described in the comments, it should never hit `new byte[]` under normal circumstances.

Reply:
The JIT isn't allowed to round up here, as it could change the behavior. Any optimizations must not change the visible behavior of the (sequential) program.

BTW: maybe 3000 bytes is a bit too big for safely stack allocating? Regular (thread) stack-sizes are big enough, but that method could be deeply nested where methods up the stack also could have larger amount of stack-space, so potentially throwing StackOverflowException? I'd cap it by some smaller values like 512, then rent from the array-pool if sizes > 512 are frequent.

For rounding up that would be 3000 -> 4096, which may be quite a difference.

But now I see that my comment isn't really complete ('cause it's too obvious for me), sorry for that.
The essence is that the value given to `stackalloc` is a constant, that is a multiple of two and not any variable length.
Perf-wise it's better to do `stackalloc byte[ConstantAsMultipleOfTwo].Slice(0, count)` than just `stackalloc byte[count]`.
That's what I tried to say with my previous comment.

The ideal pattern if count is most likely below the threshold is
```c#
Span<byte> buffer = count <= MaxStackSizeInBytes
    ? stackalloc byte[MaxStackSizeInBytes].Slice(0, count)   // Slice can be omitted when the buffer doesn't need to be exact size
    : new byte[count];
```
or w/ renting from the array-pool when larger sizes are equally likely
```c#
const int MaxStackSizeInBytes = 512;   // or whatever is a good choice here (128, 256), as long as it's a multiple of 2

byte[]? bufferFromPool = null;
Span<byte> buffer = (count <= MaxStackSizeInBytes
    ? stackalloc byte[MaxStackSizeInBytes]
    : bufferFromPool = ArrayPool<byte>.Shared.Rent(count))
    .Slice(0, count);

// Use buffer

if (bufferFromPool is not null)
{
    ArrayPool<byte>.Shared.Return(bufferFromPool);
}
```
Note: the array returned from the array-pool may also be longer than the length requested, thus it's sliced to the correct size -- except the buffer can be bigger than needed.

Reply:
Updated to suggestion